python/rmm/rmm.py

```diff
@@ -231,3 +238,48 @@ def rmm_cupy_allocator(nbytes):
     ptr = cupy.cuda.memory.MemoryPointer(mem, 0)
 
     return ptr
+
+
+def register_reinitialize_hook(func, *args, **kwargs):
+    """
+    Add a function to the list of functions ("hooks") that will be
+    called before :py:func:`~rmm.reinitialize()`.
+
+    A user or library may register hooks to perform any necessary
+    cleanup before RMM is reinitialized. For example, a library with
+    an internal cache of objects that use device memory allocated by
+    RMM can register a hook to release those references before RMM is
+    reinitialized, thus ensuring that the relevant device memory
+    resource can be deallocated.
+
+    Hooks are called in the *reverse* order they are registered.  This
+    is useful, for example, when a library registers multiple hooks
+    and needs them to run in a specific order for cleanup to be safe.
+    Hooks cannot rely on being registered in a particular order
+    relative to hooks registered by other packages, since that is
+    determined by package import ordering.
+
+    Parameters
+    ----------
+    func: callable
+        Function to be called before :py:func:`~rmm.reinitialize()`
+    args, kwargs
+        Positional and keyword arguments to be passed to `func`
+    """
+    global _reinitialize_hooks
+    _reinitialize_hooks.append((func, args, kwargs))
+    return func
+
+
+def unregister_reinitialize_hook(func):
+    """
+    Remove `func` from list of hooks that will be called before
+    :py:func:`~rmm.reinitialize()`.
+
+    If `func` was registered more than once, every instance of it will
+    be removed from the list of hooks.
+    """
+    global _reinitialize_hooks
+    _reinitialize_hooks = list(
+        filterfalse(lambda x: x[0] == func, _reinitialize_hooks)
```

Comment:
Why `filterfalse` with `==` when `filter` with `!=` would do the same thing? Then we don't need to import from itertools.
```suggestion
        filter(lambda x: x[0] != func, _reinitialize_hooks)
```

Reply:
Also we could use a comprehension, which tends to be a bit faster / more succinct.
```
_reinitialize_hooks = [x for x in _reinitialize_hooks if x[0] != func]
```

Reply:
This seems less succinct than what is already here

Reply:
The comparison I was making was:
```python
_reinitialize_hooks = list(
    filterfalse(lambda x: x[0] == func, _reinitialize_hooks)
)
```
(3 lines, 92 chars)
vs.
```python
_reinitialize_hooks = [x for x in _reinitialize_hooks if x[0] != func]
```
(1 line, 70 chars)

The difference is small. I cared more about removing the `from itertools import filterfalse` when we just needed to swap `==` for `!=`.

Reply:
Yeah the `filter` change seems fine (assuming that check behaves as expected). Tried to clarify that above